### PR TITLE
[WIP] Stack assertions

### PR DESCRIPTION
**Pull request by akrmn**
This PR defines the stack assertion construct. Inside a function, an assertion of the form `{A B C}` is equivalent to the identity function at the type `A B C -- A B C`.

I'm marking this as WIP since I'm not sure how to handle scoped type variables.

### CODE REVIEW

File: src/mirth/elab.mth [hunk 1]
```diff
@@ -299,6 +301,14 @@ def(elab-arrow-step-word!, Arrow Word -- Arrow +Elab,
     elab-token@ token-args-0 # high order not yet supported
     elab-word-sig! elab-arrow-apply-sig-type!)
 
+def(elab-arrow-step-stack-assertion!, Arrow Token -- Arrow +Elab,
+    token-brace-arg
+    dip(elab-ctx@)
```
Comment (akrmn) [inline src/mirth/elab.mth:306]: not sure if this gives the "right" `Ctx`

File: src/mirth/elab.mth [hunk 2]
```diff
@@ -299,6 +301,14 @@ def(elab-arrow-step-word!, Arrow Word -- Arrow +Elab,
...
+    elab-type-stack! drop nip
+    dup morphism-type-new! TMorphism
+    elab-arrow-apply-sig-type!
```
Comment (akrmn) [inline src/mirth/elab.mth:309]: this felt like cheating :)

Comment (sofiafaro) [reply src/mirth/elab.mth:309]: This is really good!

File: src/mirth/elab.mth [hunk 3]
```diff
@@ -299,6 +301,14 @@ def(elab-arrow-step-word!, Arrow Word -- Arrow +Elab,
     elab-token@ token-args-0 # high order not yet supported
     elab-word-sig! elab-arrow-apply-sig-type!)
 
+def(elab-arrow-step-stack-assertion!, Arrow Token -- Arrow +Elab,
+    token-brace-arg
+    dip(elab-ctx@)
+    elab-type-stack! drop nip
```
Comment (sofiafaro) [inline src/mirth/elab.mth:307]: I think this is very close. One thing we probably want to check here is that the context you get is the same as the original context `elab-ctx@`. That way you can't introduce new type variables in the assertion by accident. Can do this in a separate PR though -- it's not critical.

File: src/mirth/elab.mth (hunk 2 above)
Comment (sofiafaro) [reply src/mirth/elab.mth:309]: Oh actually I just realised this probably doesn't work . . .

Because IIRC `elab-arrow-apply-sig-type!` will freshen the morphism type before doing matching, so it's going to interact weirdly with type variables in the signature.

Probably we are better off doing some sort of manual matching here, not sure.

**Comment by sofiafaro** [comment]
Merging for now -- the issues can be fixed separately.
